cluster/service/cluster_service.go

```diff
@@ -64,6 +71,99 @@ func (c clusterService) CreateOrSaveClusterFromConfig(ctx context.Context) error
 	return nil
 }
 
+// CreateOrSaveCluster creates clusters or save updated cluster info
+func (c clusterService) CreateOrSaveCluster(ctx context.Context, clustr *repository.Cluster) error {
+	err := validate(clustr)
+	if err != nil {
+		return errs.Wrapf(err, "failed to create or save cluster named '%s'", clustr.Name)
+	}
+	return c.ExecuteInTransaction(func() error {
+		return c.Repositories().Clusters().CreateOrSave(ctx, clustr)
+	})
+}
+
+const (
+	errEmptyFieldMsg           = "empty field '%s' is not allowed"
+	errInvalidURLMsg           = "'%s' URL is invalid: '%s'"
+	errInvalidURLGenerationMsg = "unable to generate '%s' URL from '%s' (expected an 'api' subdomain)"
+	errInvalidTypeMsg          = "invalid type of cluster: '%s' (expected 'OSO', 'OCP' or 'OSD')"
+)
+
+// validate checks if all data in the given cluster is valid, and fills the missing/optional URLs using the `APIURL`
+func validate(clustr *repository.Cluster) error {
+	if clustr.Name == "" {
```

Comment:
There is no need to validate empty params if you are mentioning it as required in `design`, because goa will handle it e.g.
`{"errors":[{"code":"bad_request","detail":"[zzwRTz/z] 400 invalid_request: attribute \"name\" of request is missing and required;  parent: request","id":"CwNtrPHC","status":"400","title":"Bad Request"}]}
`

Reply:
are you sure about that ? you're showing a JSON-API response, but I'm not sure Goa returns such a thing by default. Anyways, I'll double-check that

Reply:
yes, I am sure about it. Maybe we can check by trimming fields, to catch by some weired request with non-empty space, like `"  "`

Reply:
good point, indeed. But what about the response format?

Reply:
addressed in 133b68b